### PR TITLE
query tagging behavior option

### PR DESCRIPTION
**Pull request by jgeorgie**

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/Session.cs [hunk 1]
```diff
@@ -597,7 +597,14 @@ internal Session(Domain domain, StorageNode selectedStorageNode, SessionConfigur
       SystemQuery = Query = new QueryEndpoint(new QueryProvider(this));
     }
 
-    public TagScope Tag(string tag) => new TagScope(tags ??= new List<string>(), tag);
+    public TagScope Tag(string tag)
+    {
+      tags = tags == null || Domain.Configuration.TaggingBehavior == TaggingBehavior.LastTagOverrides
+        ? new List<string>()
```
Comment (SergeiPavlov) [inline Orm/Xtensive.Orm/Orm/Session.cs:603]:
```suggestion
        ? new List<string>(1)
```

File: Version.props [hunk 2]
```diff
@@ -2,8 +2,8 @@
 <Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
 <PropertyGroup>
-    <DoVersion>7.1.0</DoVersion>
-    <DoVersionSuffix>Beta-2-dev</DoVersionSuffix>
+    <DoVersion>7.1.0.64</DoVersion>
```
Comment (SergeiPavlov) [inline Version.props:5]:
```suggestion
```

File: Version.props [hunk 3]
```diff
@@ -2,8 +2,8 @@
...
+    <DoVersionSuffix>servicetitan</DoVersionSuffix>
```
Comment (SergeiPavlov) [inline Version.props:6]:
```suggestion
```

Comment (SergeiPavlov) [inline Version.props:6]:
```suggestion
    <DoVersion>7.1.0</DoVersion>
    <DoVersionSuffix>Beta-2-dev</DoVersionSuffix>
```

**Comment by alex-kulakov** [comment]
Hello. What is the option? Can you write some description of why we need it, how it supposes to work, etc?

File: Orm/Xtensive.Orm/Orm/Configuration/DomainConfiguration.cs [hunk 4]
```diff
@@ -732,7 +742,7 @@ protected override void CopyFrom(ConfigurationBase source)
       ignoreRules = (IgnoreRuleCollection) configuration.IgnoreRules.Clone();
       shareStorageSchemaOverNodes = configuration.ShareStorageSchemaOverNodes;
       versioningConvention = (VersioningConvention) configuration.VersioningConvention.Clone();
-      
+      taggingBehavior = configuration.taggingBehavior;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Configuration/DomainConfiguration.cs:745]: Use property to be inline with already existing code.

File: Orm/Xtensive.Orm/Orm/Configuration/TaggingBehavior.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright (C) 2007-2022 Xtensive LLC.
+// This code is distributed under MIT license terms.
+// See the License.txt file in the project root for more information.
+// Created by: Georgi Sargsyan
+// Created:    2022.07.19
+
+namespace Xtensive.Orm.Configuration
+{
+  public enum TaggingBehavior
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Configuration/TaggingBehavior.cs:9]: Would be easier to understand your idea if this enum had description for default and alternative behavior

File: Orm/Xtensive.Orm/Orm/Domain.cs [hunk 6]
```diff
@@ -370,7 +370,8 @@ internal async Task<Session> OpenSessionInternalAsync(SessionConfiguration confi
 
     /// <inheritdoc/>
     public IExtensionCollection Extensions { get; private set; }
-    
+    public TaggingBehavior TaggingBehavior { get; internal set; }
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Domain.cs:373]: Why? The setting is already in the configuration so use the domain configuration setting it.

**Comment by alex-kulakov** [comment]
Additionally, If we have some plans to expand tagging settings it would be reasonable to group them under ```DomainConfiguration.QueryTagging``` or something similar, like DomainConfiguration.VersioningConvention, and all the setting connected with tagging will be there.

File: Orm/Xtensive.Orm/Orm/Configuration/TaggingBehavior.cs (hunk 5 above)
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Orm/Configuration/TaggingBehavior.cs:9]: At least raw ideas

**Comment by alex-kulakov** [comment]
@jgeorgie, @SergeiPavlov, what do you think?

File: Orm/Xtensive.Orm/Orm/Configuration/DomainConfiguration.cs (hunk 4 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Configuration/DomainConfiguration.cs]: done

File: Orm/Xtensive.Orm/Orm/Configuration/DomainConfiguration.cs [hunk 7]
```diff
@@ -117,6 +117,7 @@ public class DomainConfiguration : ConfigurationBase
     private IgnoreRuleCollection ignoreRules = new IgnoreRuleCollection();
     private VersioningConvention versioningConvention = new VersioningConvention();
     private TagsLocation tagsLocation = TagsLocation.Default;
+    private TaggingBehavior taggingBehavior = TaggingBehavior.LastTagOverrides;
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Configuration/DomainConfiguration.cs]: I agree. done

File: Orm/Xtensive.Orm/Orm/Domain.cs (hunk 6 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Domain.cs]: Simplified it

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs [hunk 8]
```diff
@@ -32,6 +33,9 @@ internal sealed partial class Translator : QueryableVisitor
 
     private readonly TranslatorContext context;
     private readonly bool tagsEnabled;
+    private readonly TaggingBehavior taggingBehavior;
+
+    private bool isAlreadyTagged;
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs:38]: Moved it into `TranslatorState.IsAlreadyTagged` flag

**Comment by SergeiPavlov** [comment]
@Mee7891 could you please review it, as you are last who touched the Tags feature

File: Orm/Xtensive.Orm/Orm/Domain.cs (hunk 6 above)
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Orm/Domain.cs]: Not so much places use it as I see. Also allowing this property to be changed during Domain instance life is inappropriate.
Do you track it somehow in ST?

**Comment by alex-kulakov** [comment]
Wait, Can't we just add parameter to ```Tag()```? Like ```Tag(string tag, bool withOverride)```

**Comment by alex-kulakov** [comment]
To be honest, I'm sick and tired of guessing what are your use-cases for this, this PR is the only way to communicate on this suggestion. Several of my requests or hints to tell me what it does, how it meant to work and what problems of current tagging it addresses were ignored. I'm close to reject it if we have no conversation.

**Comment by SergeiPavlov** [comment]
I think it is better to reject and don't merge it into the Upstream DO repository as well.

**Comment by alex-kulakov** [comment]
Rejected and closed.

**Pull request closed without merging**